README.md
# socket_io_client_flutter

Now available as a Flutter package from [pub.dev](https://pub.dev/packages/socket_io_client_flutter).

2.x.x version has sound null safety.  It can be included in pubspec.yaml as:

```yaml
dependencies:
  socket_io_client_flutter: ^2.0.0
```

Dart 3 requires sound null safety and does not support no null safety and unsound (mixed) null safety.  Flutter versions that use Dart 3 will only work with socket_io_client_flutter 2.x.x.  Flutter versions that use Dart 2.12 and above should work with socket_io_client_flutter 2.x.x, too.

1.x.x version does not have null safety and supports no null safety and unsound (mixed) null safety.  It can be included in pubspec.yaml as:

```yaml
dependencies:
  socket_io_client_flutter: ^1.0.2
```

Dart 1 and Dart 2 require no null safety.  Dart 2.12 and above supports no null safety and both unsound (mixed) and sound null safety.  Some earlier Flutter versions that use Dart 2.12 and above run socket_io_client_flutter 1.x.x automatically but more recent Flutter versions that use Dart 2.12 and above require "--no-sound-null-safety" argument to run socket_io_client_flutter 1.x.x.

Original release comment: Port socket.io-client-dart to Flutter without null safety, compatible with Flutter 1.22.6 (no null safety) and Flutter 3.0.2 (supports mixed null safety).  Add harmless long-polling extensions for xibbit.

In 1.x.x and 2.x.x versions, it is included in code as:

```dart
import 'package:socket_io_client_flutter/socket_io_client_flutter.dart';
```

# socket.io-client-dart

Port of awesome JavaScript Node.js library - [Socket.io-client v2.0.1~v3.0.3](https://github.com/socketio/socket.io-client) - in Dart

### Version info:

| socket.io-client-dart | Socket.io Server
-------------------|----------------
`v0.9.*` ~ `v1.* ` | `v2.*`
`v2.*`             | `v3.*` & `v4.*`

## Usage

**Dart Server**
```dart
import 'package:socket_io/socket_io.dart';

main() {
  // Dart server
  var io = Server();
  var nsp = io.of('/some');
  nsp.on('connection', (client) {
    print('connection /some');
    client.on('msg', (data) {
      print('data from /some => $data');
      client.emit('fromServer', "ok 2");
    });
  });
  io.on('connection', (client) {
    print('connection default namespace');
    client.on('msg', (data) {
      print('data from default => $data');
      client.emit('fromServer', "ok");
    });
  });
  io.listen(3000);
}
```
**Dart Client**
```dart

import 'package:socket_io_client/socket_io_client.dart' as IO;

main() {
  // Dart client
  IO.Socket socket = IO.io('http://localhost:3000');
  socket.onConnect((_) {
    print('connect');
    socket.emit('msg', 'test');
  });
  socket.on('event', (data) => print(data));
  socket.onDisconnect((_) => print('disconnect'));
  socket.on('fromServer', (_) => print(_));
}
```

### Connect manually

To connect the socket manually, set the option `autoConnect: false` and call `.connect()`.

For example,

<pre>
Socket socket = io('http://localhost:3000', 
    OptionBuilder()
      .setTransports(['websocket']) // for Flutter or Dart VM
      .<b>disableAutoConnect()</b>  // disable auto-connection
      .setExtraHeaders({'foo': 'bar'}) // optional
      .build()
  );
<b>socket.connect();</b>
</pre>

Note that `.connect()` should not be called if `autoConnect: true` 
(by default, it's enabled to true), as this will cause all event handlers to get registered/fired twice. See [Issue #33](https://github.com/rikulo/socket.io-client-dart/issues/33).

### Update the extra headers

```dart
Socket socket = ... // Create socket.
socket.io.options['extraHeaders'] = {'foo': 'bar'}; // Update the extra headers.
socket.io..disconnect()..connect(); // Reconnect the socket manually.
```

### Emit with acknowledgement

```dart
Socket socket = ... // Create socket.
socket.onConnect((_) {
    print('connect');
    socket.emitWithAck('msg', 'init', ack: (data) {
        print('ack $data') ;
        if (data != null) {
          print('from server $data');
        } else {
          print("Null") ;
        }
    });
});
```

### Socket connection events

These events can be listened on.

```dart
const List EVENTS = [
  'connect',
  'connect_error',
  'connect_timeout',
  'connecting',
  'disconnect',
  'error',
  'reconnect',
  'reconnect_attempt',
  'reconnect_failed',
  'reconnect_error',
  'reconnecting',
  'ping',
  'pong'
];

// Replace 'onConnect' with any of the above events.
socket.onConnect((_) {
    print('connect');
});
```

### Acknowledge with the socket server that an event has been received.

```dart
socket.on('eventName', (data) {
    final dataList = data as List;
    final ack = dataList.last as Function;
    ack(null);
});
```

## Usage (Flutter)

In Flutter env. not (Flutter Web env.) it only works with `dart:io` websocket,
 not with `dart:html` websocket or Ajax (XHR), so in this case
you have to add `setTransports(['websocket'])` when creates the socket instance.

For example,

```dart
IO.Socket socket = IO.io('http://localhost:3000',
  OptionBuilder()
      .setTransports(['websocket']) // for Flutter or Dart VM
      .setExtraHeaders({'foo': 'bar'}) // optional
      .build());
```

## Usage with stream and streambuilder in Flutter

```dart
import 'dart:async';


// STEP1:  Stream setup
class StreamSocket{
  final _socketResponse= StreamController<String>();

  void Function(String) get addResponse => _socketResponse.sink.add;

  Stream<String> get getResponse => _socketResponse.stream;

  void dispose(){
    _socketResponse.close();
  }
}

StreamSocket streamSocket =StreamSocket();

//STEP2: Add this function in main function in main.dart file and add incoming data to the stream
void connectAndListen(){
  IO.Socket socket = IO.io('http://localhost:3000',
      OptionBuilder()
       .setTransports(['websocket']).build());

    socket.onConnect((_) {
     print('connect');
     socket.emit('msg', 'test');
    });

    //When an event recieved from server, data is added to the stream
    socket.on('event', (data) => streamSocket.addResponse);
    socket.onDisconnect((_) => print('disconnect'));

}

//Step3: Build widgets with streambuilder

class BuildWithSocketStream extends StatelessWidget {
  const BuildWithSocketStream({Key key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Container(
      child: StreamBuilder(
        stream: streamSocket.getResponse ,
        builder: (BuildContext context, AsyncSnapshot<String> snapshot){
          return Container(
            child: snapshot.data,
          );
        },
      ),
    );
  }
}

```

## Troubleshooting

### Cannot connect "https" server or self-signed certificate server

- Refer to https://github.com/dart-lang/sdk/issues/34284 issue.
  The workround is to use the following code provided by [@lehno](https://github.com/lehno) on [#84](https://github.com/rikulo/socket.io-client-dart/issues/84)

```dart
class MyHttpOverrides extends HttpOverrides {
  @override
  HttpClient createHttpClient(SecurityContext? context) {
    return super.createHttpClient(context)
      ..badCertificateCallback =
          (X509Certificate cert, String host, int port) => true;
  }
}

void main() {
  HttpOverrides.global = MyHttpOverrides();
  runApp(MyApp());
}
```

### Memory leak issues in iOS when closing socket.

- Refer to https://github.com/rikulo/socket.io-client-dart/issues/108 issue.
  Please use `socket.dispose()` instead of `socket.close()` or `socket.disconnect()` to solve the memory leak issue on iOS.

### Connect_error on MacOS with SocketException: Connection failed
* Refer to https://github.com/flutter/flutter/issues/47606#issuecomment-568522318 issue.
           
By adding the following key into the to file `*.entitlements` under directory `macos/Runner/`
```
<key>com.apple.security.network.client</key>
<true/>
```

For more details, please take a look at https://flutter.dev/desktop#setting-up-entitlements

### Can't connect socket server on Flutter with Insecure HTTP connection
* Refer to https://flutter.dev/docs/release/breaking-changes/network-policy-ios-android

The HTTP connections are disabled by default on iOS and Android, so here is a workaround to this issue,
which mentioned on [stack overflow](https://stackoverflow.com/a/65730723)


## Notes to Contributors

### Fork socket.io-client-dart

If you'd like to contribute back to the core, you can [fork this repository](https://help.github.com/articles/fork-a-repo) and send us a pull request, when it is ready.

If you are new to Git or GitHub, please read [this guide](https://help.github.com/) first.

## Who Uses

- [Quire](https://quire.io) - a simple, collaborative, multi-level task management tool.
- [KEIKAI](https://keikai.io/) - a web spreadsheet for Big Data.

## Socket.io Dart Server

- [socket.io-dart](https://github.com/rikulo/socket.io-dart)

## Contributors

- Thanks [@felangel](https://github.com/felangel) for https://github.com/rikulo/socket.io-client-dart/issues/7
- Thanks [@Oskang09](https://github.com/Oskang09) for https://github.com/rikulo/socket.io-client-dart/issues/21
- Thanks [@bruce3x](https://github.com/bruce3x) for https://github.com/rikulo/socket.io-client-dart/issues/25
- Thanks [@Kavantix](https://github.com/Kavantix) for https://github.com/rikulo/socket.io-client-dart/issues/26
- Thanks [@luandnguyen](https://github.com/luandnguyen) for https://github.com/rikulo/socket.io-client-dart/issues/59
- Thanks [@jorgefspereira](https://github.com/jorgefspereira) for https://github.com/rikulo/socket.io-client-dart/pull/177
- Thanks [@fzyzcjy](https://github.com/fzyzcjy) for https://github.com/rikulo/socket.io-client-dart/pull/188
- Thanks [@darwin-morocho](https://github.com/darwin-morocho) for https://github.com/rikulo/socket.io-client-dart/pull/189
- Thanks [@chatziko](https://github.com/chatziko) for https://github.com/rikulo/socket.io-client-dart/pull/237


CHANGELOG.md
## 2.0.0

**New Feature:**

* Port null-safe socket.io-client-dart 2.0.0 to null-safe socket_io_client_flutter 2.0.0.  Port socket_io_common 1.x.x EIO=3 PacketParser to socket_io_client_flutter as PacketParser3 to replace socket_io_common 2.x.x EIO=4 PacketParser.

## 1.2.0+1

**Bug fix:**

* Narrow socket_io_common version in pubspec.yaml.

## 1.2.0

**Bug fix:**

* Update no null-safety socket_io_client_flutter with latest socket.io-client-dart commits.

## 2.0.0

**New Feature:**

* [#237](https://github.com/rikulo/socket.io-client-dart/pull/237) Allow sending an ack with multiple data items (making it consistent with emit)

## 1.0.2

**New Feature:**

* [#237](https://github.com/rikulo/socket.io-client-dart/pull/237) Allow sending an ack with multiple data items (making it consistent with emit)

## 1.0.1

**Bug fix:**

* [#188](https://github.com/rikulo/socket.io-client-dart/pull/188) Fixbug for Backoff when many attempts: "UnsupportedError: Unsupported operation: Infinity or NaN toInt"

## 2.0.0-beta.4-nullsafety.0

**New Feature:**

* [#177](https://github.com/rikulo/socket.io-client-dart/pull/177) Send credentials with the auth option

**Bug fix:**

* [#172](https://github.com/rikulo/socket.io-client-dart/issues/172) socket id's not synced

## 2.0.0-beta.3-nullsafety.0

**New Feature:**

* [#163](https://github.com/rikulo/socket.io-client-dart/issues/163) Null safety support for 2.0.0-beta

## 2.0.0-beta.3

**Bug fix:**

* [#150](https://github.com/rikulo/socket.io-client-dart/issues/150) Problem with setQuery in socket io version 3.0

## 2.0.0-beta.2

**Bug fix:**

* [#140](https://github.com/rikulo/socket.io-client-dart/issues/140) getting Error on emitWithAck() in v2 beta

## 2.0.0-beta.1

**New Feature:**

* [#130](https://github.com/rikulo/socket.io-client-dart/issues/130) Cannot connect to socket.io V3
* [#106](https://github.com/rikulo/socket.io-client-dart/issues/106) Can we combine emitWithBinary to emit?

## 1.0.0

* [#172](https://github.com/rikulo/socket.io-client-dart/issues/172) socket id's not synced

## 2.0.0-beta.3-nullsafety.0

**New Feature:**

* [#163](https://github.com/rikulo/socket.io-client-dart/issues/163) Null safety support for 2.0.0-beta

## 2.0.0-beta.3

**Bug fix:**

* [#150](https://github.com/rikulo/socket.io-client-dart/issues/150) Problem with setQuery in socket io version 3.0

## 2.0.0-beta.2

**Bug fix:**

* [#140](https://github.com/rikulo/socket.io-client-dart/issues/140) getting Error on emitWithAck() in v2 beta

## 2.0.0-beta.1

**New Feature:**

* [#130](https://github.com/rikulo/socket.io-client-dart/issues/130) Cannot connect to socket.io V3
* [#106](https://github.com/rikulo/socket.io-client-dart/issues/106) Can we combine emitWithBinary to emit?

**New Feature:**

* [#132](https://github.com/rikulo/socket.io-client-dart/issues/132) Migrating to null safety for Dart

## 1.0.2+2

**Bug fix:**

* Widen supported versions for socket_io_common, http and logging dependencies.

## 1.0.2+1

**New Feature:**

* Port socket.io-client-dart 1.0.2 to Flutter without null safety, compatible with Flutter 1.22.6 (no null safety) and Flutter 3.0.2 (supports mixed null safety).  Add harmless long-polling extensions for xibbit.

## 1.0.2

**New Feature:**

* [#237](https://github.com/rikulo/socket.io-client-dart/pull/237) Allow sending an ack with multiple data items (making it consistent with emit)

## 1.0.1

**Bug fix:**

* [#188](https://github.com/rikulo/socket.io-client-dart/pull/188) Fixbug for Backoff when many attempts: "UnsupportedError: Unsupported operation: Infinity or NaN toInt"

## 1.0.0

**New Feature:**

* [#48](https://github.com/DanielWHoward/socket_io_client_flutter/issues/1) change code for Flutter

## 1.0.0-nullsafety.0

**New Feature:**

* [#132](https://github.com/rikulo/socket.io-client-dart/issues/132) Migrating to null safety for Dart


## 0.9.12

**New Feature:**

* [#46](https://github.com/rikulo/socket.io-client-dart/issues/46) Make this library more "Darty"

## 0.9.11

**New Feature:**

* [#108](https://github.com/rikulo/socket.io-client-dart/issues/108) Need dispose method for clearing resources

## 0.9.10+2

* Fix dart analyzer warning for formatting issues.

## 0.9.10+1

* Fix dart analyzer warning.

## 0.9.10

**Bug fix:**

* [#72](https://github.com/rikulo/socket.io-client-dart/issues/72) Can't send Iterable as one packet

## 0.9.9

**Bug fix:**

* [#67](https://github.com/rikulo/socket.io-client-dart/issues/67) Retry connection backoff after 54 tries reconnections every 0 second

## 0.9.8

**Bug fix:**

* [#33](https://github.com/rikulo/socket.io-client-dart/issues/33) socket.on('receiveMessage',(data)=>print("data")) called twice


## 0.9.7+2

**New Feature:**

* [#48](https://github.com/rikulo/socket.io-client-dart/issues/48) add links to github repo in pubspec.yaml


## 0.9.7+1

**New Feature:**

* [#38](https://github.com/rikulo/socket.io-client-dart/issues/38) Improve pub.dev score


## 0.9.6+3

**Bug fix:**

* [#42](https://github.com/rikulo/socket.io-client-dart/issues/42) Error when using emitWithAck

## 0.9.5

**New Feature:**

* [#34](https://github.com/rikulo/socket.io-client-dart/issues/34) Add support for extraHeaders

**Bug fix:**

* [#39](https://github.com/rikulo/socket.io-client-dart/issues/39) The factor of Backoff with 54 retries causes an overflow


lib/src/manager.dart
/// Copyright (C) 2017 Potix Corporation. All Rights Reserved
/// History: 2017-04-26 15:27
/// Author: jumperchen<jumperchen@potix.com>
import 'dart:async';
import 'dart:math' as math;

import 'package:logging/logging.dart';
import 'package:socket_io_common/src/util/event_emitter.dart';
import 'package:socket_io_common/src/parser/parser.dart';
import '../src/on.dart';
import '../src/socket.dart';
import '../src/engine/socket.dart' as engine_socket;
import '../src/on.dart' as util;

final Logger _logger = Logger('socket_io_client_flutter:Manager');

///
/// `Manager` constructor.
///
/// @param {String} engine instance or engine uri/opts
/// @param {Object} options
/// @api public
///
class Manager extends EventEmitter {
  // Namespaces
  Map<String, Socket> nsps = {};
  List subs = [];
  Map options;

  ///
  /// Sets the `reconnection` config.
  ///
  /// @param {Boolean} true/false if it should automatically reconnect
  /// @return {Manager} self or value
  /// @api public
  ///
  bool _reconnection;

  ///
  /// Sets the reconnection attempts config.
  ///
  /// @param {Number} max reconnection attempts before giving up
  /// @return {Manager} self or value
  /// @api public
  ///
  num _reconnectionAttempts;

  ///
  /// Sets the delay between reconnections.
  ///
  /// @param {Number} delay
  /// @return {Manager} self or value
  /// @api public
  ///
  num _reconnectionDelay;
  num _randomizationFactor;
  num _reconnectionDelayMax;

  ///
  /// Sets the connection timeout. `false` to disable
  ///
  /// @return {Manager} self or value
  /// @api public
  ///
  num timeout;
  _Backoff backoff;
  String readyState = 'closed';
  String uri;
  List connecting = [];
  num lastPing;
  bool encoding = false;
  List packetBuffer = [];
  bool reconnecting = false;

  engine_socket.Socket engine;
  Encoder encoder = Encoder();
  Decoder decoder = Decoder();
  bool autoConnect;
  bool skipReconnect;

  Manager({uri, Map options}) {
    options = options ?? <dynamic, dynamic>{};

    options['path'] ??= '/socket.io';
    this.options = options;
    _reconnection = options['reconnection'] != false;
    _reconnectionAttempts = options['reconnectionAttempts'] ?? double.infinity;
    _reconnectionDelay = options['reconnectionDelay'] ?? 1000;
    reconnectionDelayMax = options['reconnectionDelayMax'] ?? 5000;
    randomizationFactor = options['randomizationFactor'] ?? 0.5;
    backoff = _Backoff(
        min: _reconnectionDelay,
        max: reconnectionDelayMax,
        jitter: randomizationFactor);
    timeout = options['timeout'] ?? 20000;
    this.uri = uri;
    autoConnect = options['autoConnect'] != false;
    if (autoConnect) open();
  }

  ///
  /// Propagate given event to sockets and emit on `this`
  ///
  /// @api private
  ///
  void emitAll(String event, [data]) {
    emit(event, data);
    for (var nsp in nsps.keys) {
      nsps[nsp].emit(event, data);
    }
  }

  ///
  /// Update `socket.id` of all sockets
  ///
  /// @api private
  ///
  void updateSocketIds() {
    for (var nsp in nsps.keys) {
      nsps[nsp].id = generateId(nsp);
    }
  }

  ///
  /// generate `socket.id` for the given `nsp`
  ///
  /// @param {String} nsp
  /// @return {String}
  /// @api private
  ///
  String generateId(String nsp) {
    if (nsp.startsWith('/')) nsp = nsp.substring(1);
    return (nsp.isEmpty ? '' : (nsp + '#')) + (engine.id ?? '');
  }

  num get randomizationFactor => _randomizationFactor;
  set randomizationFactor(num v) {
    _randomizationFactor = v;
    if (backoff != null) {
      backoff.jitter = v;
    }
  }

  ///
  /// Sets the maximum delay between reconnections.
  ///
  /// @param {Number} delay
  /// @return {Manager} self or value
  /// @api public
  ///
  num get reconnectionDelayMax => _reconnectionDelayMax;
  set reconnectionDelayMax(num v) {
    _reconnectionDelayMax = v;
    if (backoff != null) {
      backoff.max = v;
    }
  }

  ///
  /// Starts trying to reconnect if reconnection is enabled and we have not
  /// started reconnecting yet
  ///
  /// @api private
  ///
  void maybeReconnectOnOpen() {
    // Only try to reconnect if it's the first time we're connecting
    if (!reconnecting && _reconnection == true && backoff.attempts == 0) {
      // keeps reconnection from firing twice for the same reconnection loop
      reconnect();
    }
  }

  ///
  /// Sets the current transport `socket`.
  ///
  /// @param {Function} optional, callback
  /// @return {Manager} self
  /// @api public
  ///
  Manager open({callback, Map opts}) =>
      connect(callback: callback, opts: opts);

  Manager connect({callback, Map opts}) {
    _logger.fine('readyState $readyState');
    if (readyState.contains('open')) return this;

    _logger.fine('opening $uri');
    engine = engine_socket.Socket(uri, options);
    var socket = engine;
    readyState = 'opening';
    skipReconnect = false;

    // emit `open`
    var openSub = util.on(socket, 'open', (_) {
      onopen();
      if (callback != null) callback();
    });

    // emit `connect_error`
    var errorSub = util.on(socket, 'error', (data) {
      _logger.fine('connect_error');
      cleanup();
      readyState = 'closed';
      emitAll('connect_error', data);
      if (callback != null) {
        callback({'error': 'Connection error', 'data': data});
      } else {
        // Only do this if there is no fn to handle the error
        maybeReconnectOnOpen();
      }
    });

    // emit `connect_timeout`
    var timeout = this.timeout;
    if (timeout != null) {
      _logger.fine('connect attempt will timeout after $timeout');

      // set timer
      var timer = Timer(Duration(milliseconds: timeout.toInt()), () {
        _logger.fine('connect attempt timed out after $timeout');
        openSub.destroy();
        socket.close();
        socket.emit('error', 'timeout');
        emitAll('connect_timeout', timeout);
      });

      subs.add(Destroyable(() => timer.cancel()));
    }

    subs.add(openSub);
    subs.add(errorSub);

    return this;
  }

  ///
  /// Called upon transport open.
  ///
  /// @api private
  ///
  void onopen([_]) {
    _logger.fine('open');

    // clear old subs
    cleanup();

    // mark as open
    readyState = 'open';
    emit('open');

    // add subs
    var socket = engine;
    subs.add(util.on(socket, 'data', ondata));
    subs.add(util.on(socket, 'ping', onping));
    subs.add(util.on(socket, 'pong', onpong));
    subs.add(util.on(socket, 'error', onerror));
    subs.add(util.on(socket, 'close', onclose));
    subs.add(util.on(decoder, 'decoded', ondecoded));
  }

  ///
  /// Called upon a ping.
  ///
  /// @api private
  ///
  void onping([_]) {
    lastPing = DateTime.now().millisecondsSinceEpoch;
    emitAll('ping');
  }

  ///
  /// Called upon a packet.
  ///
  /// @api private
  ///
  void onpong([_]) {
    emitAll('pong', DateTime.now().millisecondsSinceEpoch - lastPing);
  }

  ///
  /// Called with data.
  ///
  /// @api private
  ///
  void ondata(data) {
    decoder.add(data);
  }

  ///
  /// Called when parser fully decodes a packet.
  ///
  /// @api private
  ///
  void ondecoded(packet) {
    emit('packet', packet);
  }

  ///
  /// Called upon socket error.
  ///
  /// @api private
  ///
  void onerror(err) {
    _logger.fine('error $err');
    emitAll('error', err);
  }

  ///
  /// Creates a socket for the given `nsp`.
  ///
  /// @return {Socket}
  /// @api public
  ///
  Socket socket(String nsp, Map opts) {
    var socket = nsps[nsp];

    var onConnecting = ([_]) {
      if (!connecting.contains(socket)) {
        connecting.add(socket);
      }
    };

    if (socket == null) {
      socket = Socket(this, nsp, opts);
      nsps[nsp] = socket;
      socket.on('connecting', onConnecting);
      socket.on('connect', (_) {
        socket.id = generateId(nsp);
      });

      if (autoConnect) {
        // manually call here since connecting event is fired before listening
        onConnecting();
      }
    }

    return socket;
  }

  ///
  /// Called upon a socket close.
  ///
  /// @param {Socket} socket
  ///
  void destroy(socket) {
    connecting.remove(socket);
    if (connecting.isNotEmpty) return;

    close();
  }

  ///
  /// Writes a packet.
  ///
  /// @param {Object} packet
  /// @api private
  ///
  void packet(Map packet) {
    _logger.fine('writing packet $packet');
    if (packet.containsKey('query') && packet['type'] == 0) {
      packet['nsp'] += '''?${packet['query']}''';
    }

    if (encoding != true) {
      // encode, then write to engine with result
      encoding = true;
      encoder.encode(packet, (encodedPackets) {
        for (var i = 0; i < encodedPackets.length; i++) {
          engine.write(encodedPackets[i], packet['options']);
        }
        encoding = false;
        processPacketQueue();
      });
    } else {
      // add packet to the queue
      packetBuffer.add(packet);
    }
  }

  ///
  /// If packet buffer is non-empty, begins encoding the
  /// next packet in line.
  ///
  /// @api private
  ///
  void processPacketQueue() {
    if (packetBuffer.isNotEmpty && encoding != true) {
      var pack = packetBuffer.removeAt(0);
      packet(pack);
    }
  }

  ///
  /// Clean up transport subscriptions and packet buffer.
  ///
  /// @api private
  ///
  void cleanup() {
    _logger.fine('cleanup');

    var subsLength = subs.length;
    for (var i = 0; i < subsLength; i++) {
      var sub = subs.removeAt(0);
      sub.destroy();
    }

    packetBuffer = [];
    encoding = false;
    lastPing = null;

    decoder.destroy();
  }

  ///
  /// Close the current socket.
  ///
  /// @api private
  ///
  void close() => disconnect();

  void disconnect() {
    _logger.fine('disconnect');
    skipReconnect = true;
    reconnecting = false;
    if ('opening' == readyState) {
      // `onclose` will not fire because
      // an open event never happened
      cleanup();
    }
    backoff.reset();
    readyState = 'closed';
    engine.close();
  }

  ///
  /// Called upon engine close.
  ///
  /// @api private
  ///
  void onclose(error) {
    _logger.fine('onclose');

    cleanup();
    backoff.reset();
    readyState = 'closed';
    emit('close', error['reason']);

    if (_reconnection == true && !skipReconnect) {
      reconnect();
    }
  }

  ///
  /// Attempt a reconnection.
  ///
  /// @api private
  ///
  Manager reconnect() {
    if (reconnecting || skipReconnect) return this;

    if (backoff.attempts >= _reconnectionAttempts) {
      _logger.fine('reconnect failed');
      backoff.reset();
      emitAll('reconnect_failed');
      reconnecting = false;
    } else {
      var delay = backoff.duration;
      _logger.fine('will wait %dms before reconnect attempt', delay);

      reconnecting = true;
      var timer = Timer(Duration(milliseconds: delay.toInt()), () {
        if (skipReconnect) return;

        _logger.fine('attempting reconnect');
        emitAll('reconnect_attempt', backoff.attempts);
        emitAll('reconnecting', backoff.attempts);

        // check again for the case socket closed in above events
        if (skipReconnect) return;

        open(callback: ([err]) {
          if (err != null) {
            _logger.fine('reconnect attempt error');
            reconnecting = false;
            reconnect();
            emitAll('reconnect_error', err['data']);
          } else {
            _logger.fine('reconnect success');
            onreconnect();
          }
        });
      });

      subs.add(Destroyable(() => timer.cancel()));
    }
    return this;
  }

  ///
  /// Called upon successful reconnect.
  ///
  /// @api private
  ///
  void onreconnect() {
    var attempt = backoff.attempts;
    reconnecting = false;
    backoff.reset();
    updateSocketIds();
    emitAll('reconnect', attempt);
  }
}

///
/// Initialize backoff timer with `opts`.
///
/// - `min` initial timeout in milliseconds [100]
/// - `max` max timeout [10000]
/// - `jitter` [0]
/// - `factor` [2]
///
/// @param {Object} opts
/// @api public
class _Backoff {
  num _ms;
  num _max;
  final num _factor;
  num _jitter;
  num attempts = 0;

  _Backoff({min = 100, max = 10000, jitter = 0, factor = 2})
      : _ms = min,
        _max = max,
        _factor = factor {
    _jitter = jitter > 0 && jitter <= 1 ? jitter : 0;
  }

  ///
  /// Return the backoff duration.
  ///
  /// @return {Number}
  /// @api public
  ///
  num get duration {
    var ms = math.min(_ms * math.pow(_factor, attempts++), 1e100);
    if (_jitter > 0) {
      var rand = math.Random().nextDouble();
      var deviation = (rand * _jitter * ms).floor();
      ms = ((rand * 10).floor() & 1) == 0 ? ms - deviation : ms + deviation;
    }
    // #39: avoid an overflow with negative value
    ms = math.min(ms, _max);
    return ms <= 0 ? _max : ms;
  }

  ///
  /// Reset the number of attempts.
  ///
  /// @api public
  ///
  void reset() {
    attempts = 0;
  }

  ///
  /// Set the minimum duration
  ///
  /// @api public
  ///
  set min(min) => _ms = min;

  ///
  /// Set the maximum duration
  ///
  /// @api public
  ///
  set max(max) => _max = max;

  ///
  /// Set the jitter
  ///
  /// @api public
  ///
  set jitter(jitter) => _jitter = jitter;
}
